### ISSUE TITLE
Magisk can't work if /system/sbin exists, but not in $PATH.

### ISSUE THREAD

**Issue by lss4** [topjohnwu/Magisk#4309]
Device: Razer Phone 2
Android version: [Palladium OS GSI arm64-bvN (Android 11)](https://sourceforge.net/projects/braiagsi/files/PalladiumOS/)
Magisk version name: fb8000b5
Magisk version code: 22104

**UPDATE & TL;DR:** If your ROM/GSI has a physical `/system/sbin` folder (or `/sbin` considering the SAR context), Magisk would default to placing itself there. However, if the ROM/GSI does not include `/system/sbin` in `$PATH`, then under normal circumstances, Magisk stuffs inside `/system/sbin` would not be directly accessible, rendering it partially non-functional.

The workaround would be to physically delete the `/system/sbin` folder via TWRP. An `rm -r` should suffice. This way, Magisk would then place itself to `/system/bin` which is directly accessible.

However, if you're unable to mount `/system` r/w (such as using a non-vndklite GSI), there's little you can do.

---
ORIGINAL POST:

It used to work with [NitrogenOS GSI](https://sourceforge.net/projects/braiagsi/files/NitrogenOS/20210405/), but not this one. Magisk is indeed installed as there are Magisk stuffs present in `/sbin`, but commands like `su` do not work. Trying to call `su` gives me `/system/bin/sh: su: inaccessible or not found`.

Flashed a few Magisk modules (NanoDroid stuffs) from TWRP and the installers all report Magisk detected and the modules work as intended.

Also, the logcat contains the following.
```
06-25 10:30:51.064     0     0 I init    : processing action (post-fs-data) from (/system/etc/init/hw/init.rc:1100)
06-25 10:30:51.064     0     0 I init    : Command 'rm /dev/.magisk_unblock' action=post-fs-data (/system/etc/init/hw/init.rc:1102) took 0ms and failed: unlink() failed: No such file or directory
06-25 10:30:51.064     0     0 I init    : starting service 'EhUGec9'...
06-25 10:30:51.126     0     0 I init    : wait for '/dev/.magisk_unblock' took 60ms
06-25 10:30:51.126     0     0 I init    : Command 'wait /dev/.magisk_unblock 40' action=post-fs-data (/system/etc/init/hw/init.rc:1104) took 60ms and succeeded
06-25 10:30:51.127     0     0 I init    : Service 'EhUGec9' (pid 1041) exited with status 0 oneshot service took 0.061000 seconds in background
06-25 10:30:51.127     0     0 I init    : Sending signal 9 to service 'EhUGec9' (pid 1041) process group...
```

Was originally thinking about this being a duplicate of #4174 but it's not really the case as `su` indeed doesn't work for whoever calling it. The core problem seems to be this:

The GSI does have `/sbin` or `/system/sbin`, but an `echo $PATH` returned me the following:
```
/product/bin:/apex/com.android.runtime/bin:/apex/com.android.art/bin:/system_ext/bin:/system/bin:/system/xbin:/odm/bin:/vendor/bin:/vendor/xbin
```
`/sbin` or `/system/sbin` was not among the listed `$PATH`, so this explains why I can't `su` despite it's indeed present. ~~If the omission of `/sbin` or `/system/sbin` is becoming a trend (since the ROMs/GSIs in question don't really have anything of their own in that folder), I think Magisk stuffs should be made to reside elsewhere or it would cease to work on more and more devices/ROMs/GSIs.~~

Another thing worth noting is that, on PalladiumOS GSI, `adb root` always works for me to get a root shell without invoking `su`. In the past, `adb root` would stop working (hang) if Magisk is used in place of phh-su, though in this case, the GSI doesn't really have phh-su installed, as it is a N flavor.

~~PS: A long time ago I tried flashing Magisk on CAOS GSI and it didn't work, even after securizing (removing phh-su) and renaming the `resetprop` (to avoid collision, which was already addressed in recent GSIs). I suspect the reason it didn't work was also probably due to the absence of `/sbin` or `/system/sbin` in `$PATH`, though I didn't pay attention to it back then.~~

~~EDIT: Don't know if it's possible to somehow patch the system to include `/sbin` or `/system/sbin` in `$PATH` if not already included...~~

**Comment by lss4** [comment]
UPDATE: Figured out the problem. I was wrong. It was the GSI's fault.

Just tried NitrogenOS again. It also doesn't have `/sbin` or `/system/sbin` in the folder in `$PATH`. However, the GSI doesn't physically have the `/system/sbin` folder so Magisk is placed inside `/system/bin` instead, which is accessible.

Restored myself back to PalladiumOS. This time, I deleted `/system/sbin` through TWRP, and now Magisk is placed inside `/system/bin` and can be correctly detected by the Manager (yay!).

So it turned out Magisk does have some workarounds, but I think the workaround might need some improvements so it would work in case the GSI does have `/system/sbin` but does not actually include that folder in `$PATH` (which caused the breakage).

**Comment by topjohnwu** [comment]
That is not something Magisk should fix. The GSI should set the proper `PATH` value

**Comment by lss4** [comment]
Okay... I understand...

I think this should be forwarded to the now-closed #3593 as an addendum, that if `/sbin` is not in `$PATH` then do not create that.

**Comment by felipecrs** [comment]
> UPDATE & TL;DR: If your ROM/GSI has a physical /system/sbin folder (or /sbin considering the SAR context), Magisk would default to placing itself there. However, if the ROM/GSI does not include /system/sbin in $PATH, then under normal circumstances, Magisk stuffs inside /system/sbin would not be directly accessible, rendering it partially non-functional.

@topjohnwu to me it seems that the magisk installer can simply verify if `/system/sbin` is on `PATH` before trying to install on it. Don't you think?
